actionpack/lib/action_controller/metal/rendering.rb

```diff
@@ -4,6 +4,17 @@ module Rendering
 
     RENDER_FORMATS_IN_PRIORITY = [:body, :text, :plain, :html]
 
+    module ClassMethods
+      # Documentation at ActionController::Renderer#render
+      delegate :render, to: :renderer
+
+      # Returns a renderer class (inherited from ActionController::Renderer)
+      # for the controller.
+      def renderer
+        @renderer ||= Renderer.for(self)
```

Comment:
`renderer` class has mutable `defaults`, but it does not matter. The reason is the same why access to the **one** `renderer` class across threads is safe even though access to underlying `defaults` is not protected by mutex.

**`defaults` stores a static (predefined) value**. That's enough to be safe in our case. 
Dynamic environment should be passed as `renderer.new(env)` (if a developer would try to use `defaults` to dynamically change the environment he would need to use a mutex himself anyway to wrap a change to `defaults` and consequent `render` call).

I can explain it further if you want.